### DISCUSSION TITLE
Extensions disappear if set insite a multiprocess pool

### DISCUSSION THREAD

**Discussion by DorianKodelja** [explosion/spaCy#8352] [Help: Coding & Implementations] [answered]
## How to reproduce the behaviour
```
import spacy
from multiprocessing import Pool
nlp = spacy.load('fr_core_news_md')

doc = nlp("les instruments c'était mieux à vent")

def tmp(doc):
    doc.set_extension('test', default=0, force=True)
    doc._.set('test', 1)
    print(f'extension inside func: {doc._.test}')
    return doc

with Pool(1) as p:
    pool_docs = p.map(tmp,[doc])
    
print(pool_docs[0]._.test)
```
The first print inside the function work as expected but the second one will trigger `AttributeError: [E046] Can't retrieve unregistered extension attribute 'test'`

If, instead of setting the extension in the pooled process, we set it before as bellow:

```
def tmp(doc):
    doc._.set('test', 1)
    print(f'extension inside func: {doc._.test}')
    return doc

doc.set_extension('test', default=0, force=True)
with Pool(1) as p:
    pool_docs = p.map(tmp,[doc])
    

print(pool_docs[0]._.test)
```

both print will work as expected.

## Your Environment
OS: tensorflow:2.2.1-gpu docker (Bionic Beaver 18.04 LTS)
- **spaCy version:** 3.0.5
- **Platform:** Linux-5.4.0-74-generic-x86_64-with-Ubuntu-18.04-bionic
- **Python version:** 3.6.9
- **Pipelines:** fr_core_news_md (3.0.0)

**Comment by adrianeboyd** [comment, marked as answer]
Extensions are currently simply registered globally and you're seeing side effects of how the global context is handled during multiprocessing.

To have everything work with multiprocessing with both fork and spawn, you would need to add the extension on init in your component (or globally in your script) and also in the method where you're accessing/adding the annotation in the extension, since your top-level script and the pipeline function won't necessarily have access to the same global context.

The second example above works in linux with fork, but if you use it with spawn (in linux by explicitly switching to spawn or in windows/osx where spawn is the default), then the custom function won't see anything about the extensions that were set globally in the main script. Some further details: https://docs.python.org/3/library/multiprocessing.html#the-spawn-and-forkserver-start-methods

If you use multiprocessing through `nlp.pipe(n_process=)` instead, it should handle these details for you by passing the custom extension context to the child processes, so the second version of your script would work without any extra code.
